Fix and test private LB support

johnsca opened the pull request:
I refactored the integration tests to cover the private LB case and uncovered an issue with how the address for the private LB was being read.

johnsca committed:
Fix and test private LB support

I refactored the integration tests to cover the private LB case and
uncovered an issue with how the address for the private LB was being
read.

johnsca commented:
Test run output:

```
$ tox -e integration
integration installed: appdirs==1.4.4,attrs==20.3.0,backcall==0.2.0,bcrypt==3.2.0,blessings==1.6,cached-property==1.5.2,certifi==2020.12.5,cffi==1.14.5,chardet==4.0.0,charm-tools==2.8.3,charmcraft==0.9.0,Cheetah3==3.2.6.post1,colander==1.7.0,cryptography==3.4.6,decorator==4.4.2,dict2colander==0.2,distlib==0.3.1,distro==1.5.0,filelock==3.0.12,httplib2==0.19.0,humanize==3.3.0,idna==2.10,iniconfig==1.1.1,ipdb==0.13.7,ipython==7.21.0,ipython-genutils==0.2.0,iso8601==0.1.14,jedi==0.18.0,Jinja2==2.11.3,jsonschema==2.5.1,juju @ https://github.com/juju/python-libjuju/archive/master.zip,jujubundlelib==0.5.6,keyring==20.0.1,launchpadlib==1.10.13,lazr.restfulclient==0.14.3,lazr.uri==1.0.5,libcharmstore==0.0.9,loadbalancer-interface==1.1.0,macaroonbakery==1.3.1,MarkupSafe==1.1.1,marshmallow==3.10.0,marshmallow-enum==1.5.1,mypy-extensions==0.4.3,oauthlib==3.1.0,ops==1.1.0,ops-reactive-interface==1.0.1,otherstuf==1.1.0,packaging==20.9,paramiko==2.7.2,parse==1.19.0,parso==0.8.1,path==15.1.2,path.py==12.5.0,pathspec==0.3.4,pbr==5.5.1,pexpect==4.8.0,pickleshare==0.7.5,pluggy==0.13.1,prompt-toolkit==3.0.18,protobuf==3.15.6,ptyprocess==0.7.0,py==1.10.0,pyasn1==0.4.8,pycparser==2.20,Pygments==2.8.1,pymacaroons==0.13.0,PyNaCl==1.4.0,pyparsing==2.4.7,pyRFC3339==1.1,pytest==6.2.2,pytest-asyncio==0.14.0,pytest-operator==0.6.0,python-dateutil==2.8.1,pytz==2021.1,PyYAML==5.4.1,requests==2.25.1,requests-toolbelt==0.9.1,requirements-parser==0.2.0,ruamel.yaml==0.15.100,SecretStorage==2.3.1,six==1.15.0,stuf==0.9.16,tabulate==0.8.7,testresources==2.0.1,theblues==0.5.2,toml==0.10.2,toposort==1.6,traitlets==5.0.5,translationstring==1.4,typing-extensions==3.7.4.3,typing-inspect==0.6.0,urllib3==1.26.4,vergit==1.0.2,virtualenv==20.4.3,wadllib==1.3.5,wcwidth==0.2.5,websockets==7.0
integration run-test-pre: PYTHONHASHSEED='3196751002'
integration run-test: commands[0] | pytest --show-capture=no --log-cli-level=INFO --disable-warnings -svv --tb native tests/integration
============================================================================================================================== test session starts ===============================================================================================================================
platform linux -- Python 3.8.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/johnsca/juju/envs/dev/charms/azure-integrator/.tox/integration/bin/python
cachedir: .tox/integration/.pytest_cache
rootdir: /home/johnsca/juju/envs/dev/charms/azure-integrator
plugins: asyncio-0.14.0, operator-0.6.0, loadbalancer-interface-1.1.0
collected 3 items                                                                                                                                                                                                                                                                

tests/integration/test_azure_integrator.py::test_build_and_deploy 
--------------------------------------------------------------------------------------------------------------------------------- live log setup ---------------------------------------------------------------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:132 Using tmp_path: /tmp/pytest-of-johnsca/pytest-12/test-azure-integrator-hax40
INFO     pytest_operator.plugin:plugin.py:179 Adding model azure:test-azure-integrator-hax4
--------------------------------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:295 Building charm azure-integrator
INFO     juju.model:model.py:1546 Deploying local:focal/azure-integrator-0
INFO     pytest_operator.plugin:plugin.py:429 Rendering charm /home/johnsca/juju/envs/dev/charms/azure-integrator/.tox/integration/lib/python3.8/site-packages/loadbalancer_interface/examples/lb-consumer
INFO     pytest_operator.plugin:plugin.py:295 Building charm lb-consumer
INFO     juju.model:model.py:1546 Deploying local:focal/lb-consumer-0
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [allocating] waiting: waiting for machine
  lb-consumer/0 [allocating] waiting: waiting for machine
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [allocating] waiting: waiting for machine
  lb-consumer/0 [allocating] waiting: waiting for machine
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [allocating] waiting: waiting for machine
  lb-consumer/0 [allocating] waiting: waiting for machine
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [allocating] waiting: waiting for machine
  lb-consumer/0 [allocating] waiting: waiting for machine
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [allocating] waiting: waiting for machine
  lb-consumer/0 [allocating] waiting: waiting for machine
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [allocating] waiting: waiting for machine
  lb-consumer/0 [allocating] waiting: waiting for machine
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [allocating] waiting: installing agent
  lb-consumer/0 [allocating] waiting: agent initializing
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] maintenance: installing charm software
  lb-consumer/0 [idle] waiting: waiting on provider
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] maintenance: installing charm software
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] maintenance: installing charm software
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] maintenance: Installing azure-cli
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] maintenance: Installing azure-cli
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] maintenance: Installing azure-cli
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] maintenance: Installing azure-cli
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
PASSED
tests/integration/test_azure_integrator.py::test_public 
--------------------------------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------------------------------
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] waiting: waiting on provider
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [idle] active: 52.149.209.82
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [idle] active: 52.149.209.82
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
PASSED
tests/integration/test_azure_integrator.py::test_private 
--------------------------------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------------------------------
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] waiting: waiting on provider
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
  lb-consumer/0 [idle] active: 192.168.0.6
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
PASSED
------------------------------------------------------------------------------------------------------------------------------- live log teardown --------------------------------------------------------------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:218 Model status:

Unit                 Machine  Status       Message
azure-integrator/0*  0        active       Ready
lb-consumer/0*       1        waiting      waiting on provider

Machine  Series      Status
0        focal       running
1        focal       running
INFO     pytest_operator.plugin:plugin.py:238 Juju error logs:

machine-1: 16:50:58 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: no container types determined
machine-0: 16:51:13 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: no container types determined

INFO     pytest_operator.plugin:plugin.py:250 Destroying machine 0
INFO     pytest_operator.plugin:plugin.py:250 Destroying machine 1
INFO     pytest_operator.plugin:plugin.py:253 Destroying model test-azure-integrator-hax4


================================================================================================================== 3 passed, 2882 warnings in 914.62s (0:15:14) ==================================================================================================================
____________________________________________________________________________________________________________________________________ summary _____________________________________________________________________________________________________________________________________
  integration: commands succeeded
  congratulations :)
```

johnsca commented:
One potential issue I saw when testing this is that if you try to reach the private LB address from the lb-consumer unit rather than the az-integrator, it hangs and eventually times out.  I'm not sure if this is a concern, but I could see it being one.  Do we maybe need a rule added for self-access?  (NB: I didn't actually test that on the public case, since I only noticed it near the very end.  Maybe I should update the test to run the curl from both units in both cases?)

Cynerva commented:
> One potential issue I saw when testing this is that if you try to reach the private LB address from the lb-consumer unit rather than the az-integrator, it hangs and eventually times out.

That's weird. I could see this being a problem depending on the consumer charm. I don't think it should be a problem for kubernetes-master since we generally configure the charm and its services to the Kubernetes API via local private IP, not via loadbalancer IPs. Not sure about vault or other potential consumers.

Cynerva commented on tests/integration/test_azure_integrator.py:
```diff
@@ -1,53 +1,84 @@
+            await lb_consumer.remove_relation("lb-provider", "azure-integrator")
+            await ops_test.model.wait_for_idle()
+    except Exception:
+        raise
```
Nitpick, this:

```
try:
    stuff1()
except Exception:
    raise
else:
    stuff2()
```

Isn't that equivalent to:

```
stuff1()
stuff2()
```

?

Cynerva commented:
Approved/LGTM but see comments.

joedborg reviewed (approved):
Looks good to me

johnsca committed:
Clean up integration test, test connectivity from units

There was a lot of duplication between the integration tests, so I
cleaned it up by using parameterization. Also addressed the review
comment about the try / else, and added a connectivity check from the
units (which is currently failing for self-access on the internal LB).

johnsca commented:
Output from updated tests:

```
tests/integration/test_azure_integrator.py::test_lb[public] 
--------------------------------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------------------------------
INFO     test_azure_integrator:test_azure_integrator.py:30 Creating public LB
INFO     test_azure_integrator:test_azure_integrator.py:35 Waiting for LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] waiting: waiting on provider
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
  lb-consumer/0 [idle] waiting: waiting on provider response
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [idle] active: 23.96.56.71
INFO     test_azure_integrator:test_azure_integrator.py:39 Verifying LB components
INFO     test_azure_integrator:test_azure_integrator.py:51 Confirming external access to http://23.96.56.71/
INFO     test_azure_integrator:test_azure_integrator.py:56 Confirming access from azure-integrator/0 to http://23.96.56.71/
INFO     test_azure_integrator:test_azure_integrator.py:56 Confirming access from lb-consumer/0 to http://23.96.56.71/
INFO     test_azure_integrator:test_azure_integrator.py:65 Cleaning up LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] active: 23.96.56.71
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
PASSED
tests/integration/test_azure_integrator.py::test_lb[internal] 
--------------------------------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------------------------------
INFO     test_azure_integrator:test_azure_integrator.py:30 Creating internal LB
INFO     test_azure_integrator:test_azure_integrator.py:35 Waiting for LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] waiting: waiting on provider
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
  lb-consumer/0 [idle] waiting: waiting on provider response
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [idle] active: 192.168.0.6
INFO     test_azure_integrator:test_azure_integrator.py:39 Verifying LB components
INFO     test_azure_integrator:test_azure_integrator.py:56 Confirming access from azure-integrator/0 to http://192.168.0.6/
INFO     test_azure_integrator:test_azure_integrator.py:56 Confirming access from lb-consumer/0 to http://192.168.0.6/
ERROR    test_azure_integrator:test_azure_integrator.py:62 Failed: assert 'nginx' in '  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dlo...0      0 --:--:--  0:02:10 --:--:--     0\ncurl: (28) Failed to connect to 192.168.0.6 port 80: Connection timed out\n'
INFO     test_azure_integrator:test_azure_integrator.py:65 Cleaning up LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [idle] active: 192.168.0.6
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
FAILED
------------------------------------------------------------------------------------------------------------------------------- live log teardown --------------------------------------------------------------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:218 Model status:

Unit                 Machine  Status       Message
lb-consumer/0*       1        waiting      waiting on provider
azure-integrator/0*  0        active       Ready

Machine  Series      Status
0        focal       running
1        focal       running
INFO     pytest_operator.plugin:plugin.py:238 Juju error logs:

machine-1: 10:59:22 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: no container types determined
machine-0: 10:59:24 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: no container types determined
unit-azure-integrator-0: 11:21:20 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer not found (not found)
unit-lb-consumer-0: 11:21:20 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator not found (not found)
unit-lb-consumer-0: 11:21:20 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator not found (not found)
unit-azure-integrator-0: 11:21:20 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer not found (not found)
unit-lb-consumer-0: 12:02:01 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator not found (not found)
unit-lb-consumer-0: 12:02:01 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator not found (not found)
unit-azure-integrator-0: 12:02:01 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer not found (not found)
unit-azure-integrator-0: 12:02:01 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer not found (not found)



==================================================================================================================================== FAILURES ====================================================================================================================================
_______________________________________________________________________________________________________________________________ test_lb[internal] ________________________________________________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/johnsca/juju/envs/dev/charms/azure-integrator/tests/integration/test_azure_integrator.py", line 59, in test_lb
    assert "nginx" in output
AssertionError: assert 'nginx' in '  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dlo...0      0 --:--:--  0:02:10 --:--:--     0\ncurl: (28) Failed to connect to 192.168.0.6 port 80: Connection timed out\n'
============================================================================================================================ short test summary info =============================================================================================================================
FAILED tests/integration/test_azure_integrator.py::test_lb[internal] - AssertionError: assert 'nginx' in '  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dlo...0      0 --:--:--  0:02:10 --:--:--     0\ncu...
```

johnsca committed:
Skip self-connectivity check for internal LB due to upstream limitation

johnsca commented:
Per doc that @Cynerva found, skipped self-connectivity check for the internal LB case.  Also removed use of branch for libjuju since 2.8.6 has been released.  Test run output:

```
tests/integration/test_azure_integrator.py::test_lb[public] 
--------------------------------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------------------------------
INFO     test_azure_integrator:test_azure_integrator.py:30 Creating public LB
INFO     test_azure_integrator:test_azure_integrator.py:35 Waiting for LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] waiting: waiting on provider
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [idle] active: 13.92.87.68
INFO     test_azure_integrator:test_azure_integrator.py:39 Verifying LB components
INFO     test_azure_integrator:test_azure_integrator.py:51 Confirming external access to http://13.92.87.68/
INFO     test_azure_integrator:test_azure_integrator.py:62 Confirming access from azure-integrator/0 to http://13.92.87.68/
INFO     test_azure_integrator:test_azure_integrator.py:62 Confirming access from lb-consumer/0 to http://13.92.87.68/
INFO     test_azure_integrator:test_azure_integrator.py:71 Cleaning up LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] active: 13.92.87.68
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
PASSED
tests/integration/test_azure_integrator.py::test_lb[internal] 
--------------------------------------------------------------------------------------------------------------------------------- live log call ----------------------------------------------------------------------------------------------------------------------------------
INFO     test_azure_integrator:test_azure_integrator.py:30 Creating internal LB
INFO     test_azure_integrator:test_azure_integrator.py:35 Waiting for LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [idle] active: Ready
  lb-consumer/0 [executing] waiting: waiting on provider
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  lb-consumer/0 [idle] active: 192.168.0.6
INFO     test_azure_integrator:test_azure_integrator.py:39 Verifying LB components
INFO     test_azure_integrator:test_azure_integrator.py:62 Confirming access from azure-integrator/0 to http://192.168.0.6/
INFO     test_azure_integrator:test_azure_integrator.py:71 Cleaning up LB
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
  lb-consumer/0 [idle] active: 192.168.0.6
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
INFO     juju.model:model.py:2257 Waiting for model:
  azure-integrator/0 [executing] active: Ready
PASSED
------------------------------------------------------------------------------------------------------------------------------- live log teardown --------------------------------------------------------------------------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:218 Model status:

Unit                 Machine  Status       Message
lb-consumer/0*       1        waiting      waiting on provider
azure-integrator/0*  0        active       Ready

Machine  Series      Status
0        focal       running
1        focal       running
INFO     pytest_operator.plugin:plugin.py:238 Juju error logs:

machine-1: 10:59:22 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: no container types determined
machine-0: 10:59:24 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: no container types determined
unit-azure-integrator-0: 11:21:20 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer not found (not found)
unit-lb-consumer-0: 11:21:20 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator not found (not found)
unit-lb-consumer-0: 11:21:20 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#azure-integrator not found (not found)
unit-azure-integrator-0: 11:21:20 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#3#lb-consumer not found (not found)
unit-lb-consumer-0: 12:02:01 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator not found (not found)
unit-lb-consumer-0: 12:02:01 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#azure-integrator not found (not found)
unit-azure-integrator-0: 12:02:01 ERROR juju.api.watcher error trying to stop watcher: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer not found (not found)
unit-azure-integrator-0: 12:02:01 ERROR juju.worker.dependency "uniter" manifold worker returned unexpected error: relation scope id "406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer": settings 406b58ac-7fc4-4509-8170-4d527fa9c188:r#7#lb-consumer not found (not found)
```
